File: leetcode/890.查找和替换模式.go
package main

import "fmt"

/*
 * @lc app=leetcode.cn id=890 lang=golang
 *
 * [890] 查找和替换模式
 */

// @lc code=start
func findAndReplacePattern(words []string, pattern string) []string {
	res := make([]string, 0)
	for _, word := range words {
		if !check(word, pattern) {
			continue
		}
		res = append(res, word)
	}
	return res
}

func check(word string, pattern string) bool {
	cache := make(map[byte]byte)
	cache2 := make(map[byte]byte)
	for i, _ := range pattern {
		if value, ok := cache[pattern[i]]; ok && value != word[i] {
			return false
		}
		if value, ok := cache2[word[i]]; ok && value != pattern[i] {
			return false
		}
		cache[pattern[i]] = word[i]
		cache2[word[i]] = pattern[i]
	}
	return true
}

// @lc code=end
func main() {
	fmt.Println(findAndReplacePattern([]string{"abc", "deq", "mee", "aqq", "dkd", "ccc"}, "abb"))
}

File: leetcode/890.查找和替换模式.py
#
# @lc app=leetcode.cn id=890 lang=python3
#
# [890] 查找和替换模式
#
from typing import List


# @lc code=start
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        n = len(pattern)

        def check(word: str) -> bool:
            cache = {}
            cache2 = {}
            for i in range(n):
                if pattern[i] in cache and cache[pattern[i]] != word[i]:
                    return False
                if word[i] in cache2 and cache2[word[i]] != pattern[i]:
                    return False
                cache[pattern[i]] = word[i]
                cache2[word[i]] = pattern[i]
            return True

        return [word for word in words if check(word)]


# @lc code=end
print(Solution().findAndReplacePattern(
    words=["abc", "deq", "mee", "aqq", "dkd", "ccc"],
    pattern="abb"
))
